src/pyrokinetics/diagnostics/saturation_rules.py

```diff
@@ -0,0 +1,247 @@
+            growth_rate_tolerance < gamma_tolerance, 0.0
+        )
+
+        heat_tot = data["heat"].sum(dim=("field", "species"))
+        heat = (
+            data["heat"]
+            .where(growth_rate_tolerance < gamma_tolerance, 0.0)
+            .sum(dim="field")
+            / heat_tot
+        )
+        particle = (
+            data["particle"]
+            .where(growth_rate_tolerance < gamma_tolerance, 0.0)
+            .sum(dim="field")
+            / heat_tot
+        )
+
+        field_squared = (
+            np.abs(eigenfunctions.where(growth_rate_tolerance < gamma_tolerance, 0.0))
+            ** 2
+        )
+
+        # Set up Jacobian and k_perp
+        k_perp = field_squared.data * 0.0
+        jacobian = field_squared.data * 0.0
+        field_factor = field_squared.data * 0.0
+
+        # Need to load MetricTerms
+        pyro.load_metric_terms(ntheta=1024)
+        theta_metric = pyro.metric_terms.regulartheta
+
+        # Add one to cover additional theta grids (like in CGYRO)
+        nperiod = pyro.numerics.nperiod + 1
+        g_tt = pyro.metric_terms.field_aligned_covariant_metric("theta", "theta")
+
+        # Can remove when we adjust outputs to handle equal_arc = True in GS2
+        if equal_arc_theta:
+            # Parallel gradient
+            grho = np.sqrt(g_tt)
+
+            # regulartheta foes from -pi to pi
+            l_theta = cumulative_trapezoid(grho, theta_metric, initial=0.0)
+            g_tt_eq = (l_theta[-1] / (2 * np.pi)) ** 2
+
+            l_theta *= 1.0 / l_theta[-1] * 2 * np.pi
+            l_theta += -np.pi
+        else:
+            l_theta = theta_metric
+
+        # Geometric theta on grid on evenly spaced l(theta) grid
+        even_l_theta = np.linspace(-np.pi, np.pi, len(l_theta))
+        theta_geo = np.interp(even_l_theta, l_theta, theta_metric)
+
+        m = np.linspace(-(nperiod - 1), nperiod - 1, 2 * nperiod - 1)
+        ntheta = len(theta_geo) - 1
+        m = np.repeat(m, ntheta)
+        theta_geo_long = np.tile(theta_geo[:-1], 2 * nperiod - 1) + 2.0 * np.pi * m
+
+        # L(theta) grid on regular theta grid long
+        m = np.linspace(-(nperiod - 1), nperiod - 1, 2 * nperiod - 1)
+        ntheta = len(l_theta) - 1
+        m = np.repeat(m, ntheta)
+        l_theta_long = np.tile(l_theta[:-1], 2 * nperiod - 1) + 2.0 * np.pi * m
+
+        for itheta0, theta0 in enumerate(theta0s.data):
+            theta_long, k_perp_long = pyro.metric_terms.k_perp(
+                ky=1.0, theta0=theta0, nperiod=nperiod
+            )
+
+            # Interp on to equal arc grid
+            if equal_arc_theta:
+                k_perp_interp = np.interp(theta, l_theta_long, k_perp_long)
+            else:
+                k_perp_interp = np.interp(theta, theta_long, k_perp_long)
+            for iky, ky in enumerate(kys.data):
+                # Technically k_perp / ky
+                k_perp[iky, itheta0, :, :] = k_perp_interp.m * ky * bunit_over_b0
```

Comment:
Why is `k_perp_interp` multiplied by `bunit_over_b0`? Is it required to convert it into pyro units?

Reply:
Its more to do with the defn of `k_perp` and to match the thing used in GS2/CGYRO we need it. We could put it inside the actual `k_perp` method. I would need to delve into the exact way we have defined `k_perp` in our diagnostic but this is a little confusing I agree

Reply:
To elaborate:

In our CI testing, when we compare our calculated k_perp to the GS2 value we need the `bunit_over_b0`

https://github.com/pyro-kinetics/pyrokinetics/blob/unstable/tests/local_geometry/test_metric_terms.py#L199-L223

But when we compare to the `k_perp` in CGYRO we don't need the bunit_over_b0

https://github.com/gafusion/gacode/blob/29268b4b42b17cdc47376335e96bbde13e1857ee/cgyro/src/cgyro_equilibrium.F90#L260-L261

This is probably due to the definition of ky in the different codes. I have to manually add this factor into the QLGYRO implementation
https://github.com/gafusion/gacode/blob/6d8e83899c0c86fa4863f031aeeaa56eb6e0dceb/qlgyro/src/qlgyro_cgyro_cleanup.F90#L45